openfermioncirq/gates/fermionic_simulation.py

```diff
@@ -76,20 +77,119 @@ def state_swap_eigen_component(x: str, y: str, sign: int = 1, angle: float=0):
     if sign not in (-1, 1):
         raise ValueError('sign not in (-1, 1)')
 
-    dim = 2 ** len(x)
+    dim = 2**len(x)
     i, j = int(x, 2), int(y, 2)
 
     component = np.zeros((dim, dim), dtype=np.complex128)
     component[i, i] = component[j, j] = 0.5
-    component[j, i]= sign * 0.5 * 1j**(angle * 2 / np.pi)
-    component[i, j]= sign * 0.5 * 1j**(-angle * 2 / np.pi)
+    component[j, i] = sign * 0.5 * 1j**(angle * 2 / np.pi)
+    component[i, j] = sign * 0.5 * 1j**(-angle * 2 / np.pi)
     return component
 
 
-class QuadraticFermionicSimulationGate(
-        cirq.EigenGate,
-        cirq.InterchangeableQubitsGate,
-        cirq.TwoQubitGate):
+@cirq.value_equality(approximate=True)
+class ParityPreservingFermionicGate(cirq.Gate, metaclass=abc.ABCMeta):
+    r"""The Jordan-Wigner transform of :math:`\exp(-i H)` for a fermionic
+    Hamiltonian :math:`H`.
+
+    Each subclass corresponds to a set of generators :math:`\{G_i\}`
+    corresponding to the family of Hamiltonians :math:`\sum_i w_i G_i +
+    \text{h.c.}`, where the weights :math:`w_i \in \mathbb C` are specified by
+    the instance.
+
+    The Jordan-Wigner mapping maps the fermionic modes :math:`(0, \ldots, n -
+    1)` to qubits :math:`(0, \ldots, n - 1)`, respectively.
+
+    Each generator :math:`G_i` must be a linear combination of fermionic
+    monomials consisting of an even number of creation/annihilation operators.
+    This is so that the Jordan-Wigner transform acts only on the gate's qubits,
+    even when the fermionic modes are offset as part of a larger Jordan-Wigner
+    string.
+    """
+
+    def __init__(
+            self,
+            weights: Optional[Tuple[complex, ...]] = None,
+            absorb_exponent: bool = False,
+            exponent: cirq.TParamVal = 1.0,
+            global_shift: float = 0.0,
+    ) -> None:
+        """A fermionic interaction.
+
+        Args:
+            weights: The weights of the terms in the Hamiltonian.
+            absorb_exponent: Whether to absorb the given exponent into the
+                weights. If true, the exponent of the return gate is `1`.
+                Defaults to `False`.
+        """
+        if weights is None:
+            weights = (1.,) * self.num_weights
+        self.weights = weights
+
+        self._exponent = exponent
+        self._global_shift = global_shift
+        self._canonical_exponent_cached = None
+
+        if absorb_exponent:
+            self.absorb_exponent_into_weights()
+
+    @abc.abstractproperty
+    def num_weights(self) -> int:
```

Comment:
This shouldn't be abstract, since it is `len(self.fermion_generator_components)`.

Reply:
Done.